src/detection/scripts/ship_generator.py

```diff
@@ -0,0 +1,334 @@
+        min(maxYcoord - rectangle_js[0][1], maxYcoord - rectangle_js[2][1])/h,
+        max(maxYcoord - rectangle_js[0][1], maxYcoord - rectangle_js[2][1])/h
+    ]
+
+
+def match_percent_to_pixel(rect, img_w, img_h):
+    pixels = []
+    pixels.append(rect[0]*img_w)
+    pixels.append(rect[1]*img_w)
+    pixels.append(rect[2]*img_h)
+    pixels.append(rect[3]*img_h)
+
+    return [int(pixel) for pixel in pixels]
+
+
+def create_coordinate_rectangle(rectangle):
+    rectangle_js = rectangle['geometry']['coordinates'][0]
+    return [
+        min(rectangle_js[0][0], rectangle_js[1][0]),  # minX/left
+        max(rectangle_js[0][0], rectangle_js[1][0]),  # maxX/right
+        min(rectangle_js[0][1], rectangle_js[2][1]),  # minY/bottom
+        max(rectangle_js[0][1], rectangle_js[2][1])   # maxY/top
+    ]
+
+
+def shapely_polygon_map_to_grid(rect):
+    """
+    shape: Shapely box geometry
+
+    returns transformed coordinates in [minx, maxx, miny, maxy]
+    """
+    shape = geometry.geo.box(*rect)
+    project = partial(
+            pyproj.transform,
+            pyproj.Proj(init='epsg:32648'),
+            pyproj.Proj(init='epsg:32648'))
+    new_shape = transform(project, shape)
+    # Note that a box [1,2,3,4] is altered and then dumped as
+    # [(3.0, 2.0), (3.0, 4.0), (1.0, 4.0), (1.0, 2.0), (3.0, 2.0)]
+    pts = geometry.base.dump_coords(new_shape)
+
+    return [pts[3][0], pts[3][1], pts[4][0], pts[1][1]]
+
+
+def rasterio_index(rect, src):
+    ul = src.index(rect[0], rect[2])
+    lr = src.index(rect[1], rect[3])
+
+    minx = min(ul[0], lr[0])
+    maxx = max(ul[0], lr[0])
+    miny = min(ul[1], lr[1])
+    maxy = max(ul[1], lr[1])
+
+    return [minx, maxx, miny, maxy]
+
+
+def convert_to_pixels(src_ds, ogr_filename, img_corners):
+    with open(ogr_filename) as ogr:
+        ogr_dict = json.load(ogr)
+
+    coord_rectangles = [create_coordinate_rectangle(feat)
+                        for feat in ogr_dict['features'][::2]]
+    pixel_coords = [rasterio_index(rect, src_ds) for rect in coord_rectangles]
+
+    return pixel_coords
+
+
+def window_ordered_coords(bbox):
+    # ymin, ymax, xmin, xmax
+    return ((bbox[0], bbox[2]), (bbox[1], bbox[3]))
+
+
+def bbox_ordered_coords(rect):
+    # xmin, ymax, xmax, ymin
+    return rect[0], rect[2], rect[1], rect[3]
+
+
+def check_in_bbox(x, y, bbox):
+    if (x > bbox[0] and x <= bbox[1]) and (y > bbox[2] and y <= bbox[3]):
+        return True
+    return False
+
+
+def contains(big, sml):
+    if big[0] <= sml[0] and big[1] <= sml[1] and big[2] >= sml[2] and \
+       big[3] >= sml[3]:
+        return True
+    return False
+
+
+def expand_window_with_offset(bbox):
+    """
+    This commented out portion is another method of generating offset,
+    with the new window location covering a larger area
+    box_w = abs(bbox[0] - bbox[1])
+    box_h = abs(bbox[2] - bbox[3])
+    center_x = (bbox[0] + bbox[1])/2
+    center_y = (bbox[2] + bbox[3])/2
+    valid_w = 512 - box_w
```

Comment:
These constants, like 512 and 128 should be extracted to variables so we can know what they mean and can be adjusted more easily.